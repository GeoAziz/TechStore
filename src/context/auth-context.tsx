"use client";

import { createContext, useContext, useEffect, useState, ReactNode } from 'react';
import { onAuthStateChanged, User } from 'firebase/auth';
import { auth } from '@/lib/firebase';

export type UserRole = 'admin' | 'vendor' | 'client';

interface AuthContextType {
  user: User | null;
  loading: boolean;
  role: UserRole | null;
}

const AuthContext = createContext<AuthContextType>({
  user: null,
  loading: true,
  role: null,
});

export const AuthProvider = ({ children }: { children: ReactNode }) => {
  const [user, setUser] = useState<User | null>(null);
  const [loading, setLoading] = useState(true);
  const [role, setRole] = useState<UserRole | null>(null);

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (user) => {
      setUser(user);
      if (user) {
        // Simulate role assignment
        if (user.email === 'admin@zizo.net') {
          setRole('admin');
        } else if (user.email === 'vendor@zizo.net') {
          setRole('vendor');
        } else {
          setRole('client');
        }
      } else {
        setRole(null);
      }
      setLoading(false);
    });

    return () => unsubscribe();
  }, []);

  return (
    <AuthContext.Provider value={{ user, loading, role }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => useContext(AuthContext);
