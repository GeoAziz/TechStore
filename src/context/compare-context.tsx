
"use client";

import { createContext, useContext, useState, useEffect, ReactNode } from 'react';
import type { Product } from '@/lib/types';
import { useToast } from '@/hooks/use-toast';

const MAX_COMPARE_ITEMS = 4;

interface CompareContextType {
  compareItems: Product[];
  addToCompare: (product: Product) => void;
  removeFromCompare: (productId: string) => void;
  clearCompare: () => void;
}

const CompareContext = createContext<CompareContextType | undefined>(undefined);

export const CompareProvider = ({ children }: { children: ReactNode }) => {
  const [compareItems, setCompareItems] = useState<Product[]>([]);
  const { toast } = useToast();
  const [isInitialized, setIsInitialized] = useState(false);

  useEffect(() => {
    try {
      const storedItems = localStorage.getItem('compareItems');
      if (storedItems) {
        setCompareItems(JSON.parse(storedItems));
      }
    } catch (error) {
      console.error("Failed to parse compare items from localStorage", error);
      setCompareItems([]);
    }
    setIsInitialized(true);
  }, []);

  useEffect(() => {
    if (isInitialized) {
      localStorage.setItem('compareItems', JSON.stringify(compareItems));
    }
  }, [compareItems, isInitialized]);

  const addToCompare = (product: Product) => {
    if (compareItems.length >= MAX_COMPARE_ITEMS) {
      toast({
        variant: "destructive",
        title: "Compare List Full",
        description: `You can only compare up to ${MAX_COMPARE_ITEMS} items at a time.`,
      });
      return;
    }
    if (!compareItems.some(item => item.id === product.id)) {
      setCompareItems(prevItems => [...prevItems, product]);
    }
  };

  const removeFromCompare = (productId: string) => {
    setCompareItems(prevItems => prevItems.filter(item => item.id !== productId));
  };

  const clearCompare = () => {
    setCompareItems([]);
  };

  return (
    <CompareContext.Provider value={{ compareItems, addToCompare, removeFromCompare, clearCompare }}>
      {children}
    </CompareContext.Provider>
  );
};

export const useCompare = () => {
  const context = useContext(CompareContext);
  if (context === undefined) {
    throw new Error('useCompare must be used within a CompareProvider');
  }
  return context;
};
