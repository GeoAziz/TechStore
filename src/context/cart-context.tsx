
"use client";

import React, { createContext, useContext, useState, useEffect, ReactNode, useCallback } from "react";
import type { CartItem } from '@/lib/types';
import { useAuth } from './auth-context';
import { db } from '@/lib/firebase';
import { collection, onSnapshot, doc, deleteDoc } from 'firebase/firestore';
import { useToast } from '@/hooks/use-toast';
import { removeFromCart as serverRemoveFromCart } from "@/lib/firestore-service";

interface CartContextType {
  cartItems: CartItem[];
  cartCount: number;
  subtotal: number;
  total: number;
  loading: boolean;
  clearClientCart: () => void;
  removeFromCart: (productId: string) => Promise<void>;
}

const CartContext = createContext<CartContextType | undefined>(undefined);

const SHIPPING_COST = 2500;

export const CartProvider = ({ children }: { children: ReactNode }) => {
  const { user } = useAuth();
  const { toast } = useToast();
  const [cartItems, setCartItems] = useState<CartItem[]>([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    if (user) {
      setLoading(true);
      const cartRef = collection(db, 'users', user.uid, 'cart');
      const unsubscribe = onSnapshot(cartRef, (snapshot) => {
        const items = snapshot.docs.map(doc => ({ ...doc.data() } as CartItem));
        setCartItems(items);
        setLoading(false);
      }, (error) => {
        console.error("Error fetching cart:", error);
        toast({ variant: "destructive", title: "Error", description: "Could not load cart items." });
        setLoading(false);
      });

      return () => unsubscribe();
    } else {
      setCartItems([]);
      setLoading(false);
    }
  }, [user, toast]);
  
  const clearClientCart = useCallback(() => {
    setCartItems([]);
  }, []);
  
  const removeFromCart = useCallback(async (productId: string) => {
    if (!user) {
      toast({ variant: "destructive", title: "Error", description: "You must be logged in." });
      return;
    }
    await serverRemoveFromCart(user.uid, productId);
    // The onSnapshot listener will automatically update the local state.
  }, [user, toast]);

  const cartCount = cartItems.reduce((acc, item) => acc + item.quantity, 0);
  const subtotal = cartItems.reduce((acc, item) => acc + (item.price * item.quantity), 0);
  const total = subtotal > 0 ? subtotal + SHIPPING_COST : 0;

  return (
    <CartContext.Provider value={{ cartItems, cartCount, subtotal, total, loading, clearClientCart, removeFromCart }}>
      {children}
    </CartContext.Provider>
  );
};

export const useCart = () => {
  const context = useContext(CartContext);
  if (context === undefined) {
    throw new Error('useCart must be used within a CartProvider');
  }
  return context;
};
