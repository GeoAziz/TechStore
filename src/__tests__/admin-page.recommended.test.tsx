import React from 'react';
import { render, screen, fireEvent, within, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom'; // Ensure jest-dom matchers are loaded
import AdminPage from '../app/admin/page';
import { AuthProvider } from '../context/auth-context';
import { act } from 'react-dom/test-utils';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';

// Mock dependencies
jest.mock('../lib/firestore-service', () => ({
  getProducts: jest.fn(() => Promise.resolve([])),
  getOrders: jest.fn(() => Promise.resolve([])),
  getUsers: jest.fn(() => Promise.resolve([])),
  getAuditLogs: jest.fn(() => Promise.resolve([])),
  deleteProduct: jest.fn(() => Promise.resolve({ success: true })),
  deleteMultipleProducts: jest.fn(() => Promise.resolve({ success: true })),
  addProduct: jest.fn(() => Promise.resolve({ success: true })),
  updateProduct: jest.fn(() => Promise.resolve({ success: true })),
  updateOrderStatus: jest.fn(() => Promise.resolve({ success: true })),
  deleteUser: jest.fn(() => Promise.resolve({ success: true })),
  updateUserRole: jest.fn(() => Promise.resolve({ success: true })),
}));

jest.mock('../context/auth-context', () => {
  return {
    useAuth: () => ({
      user: { uid: 'admin1', email: 'admin@example.com' },
      loading: false,
      role: 'admin',
      isAddProductDialogOpen: false,
      setAddProductDialogOpen: jest.fn(),
    }),
    AuthProvider: function AuthProvider({ children }: { children: React.ReactNode }) {
      return <>{children}</>;
    },
  };
});

function customRender(ui: React.ReactElement) {
  return render(
    <MemoryRouterProvider>
      {ui}
    </MemoryRouterProvider>
  );
}

// Edge case: Empty state
it('shows empty state when no products', async () => {
  jest.spyOn(require('../lib/firestore-service'), 'getProducts').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getOrders').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getUsers').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getAuditLogs').mockResolvedValueOnce([]);
  customRender(
    <AuthProvider>
      <AdminPage defaultTab="inventory" />
    </AuthProvider>
  );
  await waitFor(() => expect(screen.queryByText(/loading/i)).not.toBeInTheDocument(), { timeout: 4000 });
  expect(await screen.findByText(/total: 0 products/i)).toBeInTheDocument();
});

// Out of stock badge
it('shows out of stock badge for products with stock 0', async () => {
  jest.spyOn(require('../lib/firestore-service'), 'getProducts').mockResolvedValueOnce([
    {
      id: 'p1',
      name: 'Test',
      imageUrl: '',
      price: 100,
      stock: 0,
      category: 'Electronics',
      brand: '',
      subcategory: '',
      isFeatured: false,
      createdAt: new Date().toISOString(),
      checked: false,
      description: 'Test product',
      updatedAt: new Date().toISOString(),
      status: 'active'
    }
  ]);
  jest.spyOn(require('../lib/firestore-service'), 'getProducts').mockResolvedValueOnce([
    {
      id: 'p1',
      name: 'Test Out of Stock',
      imageUrl: '',
      price: 100,
      stock: 0,
      category: 'Electronics',
      brand: '',
      subcategory: '',
      isFeatured: false,
      createdAt: new Date().toISOString(),
      checked: false,
      description: 'Test product',
      updatedAt: new Date().toISOString(),
      status: 'active'
    },
    {
      id: 'p2',
      name: 'Test In Stock',
      imageUrl: '',
      price: 200,
      stock: 10,
      category: 'Electronics',
      brand: '',
      subcategory: '',
      isFeatured: false,
      createdAt: new Date().toISOString(),
      checked: false,
      description: 'Test product 2',
      updatedAt: new Date().toISOString(),
      status: 'active'
    }
  ]);
  jest.spyOn(require('../lib/firestore-service'), 'getOrders').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getUsers').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getAuditLogs').mockResolvedValueOnce([]);
  customRender(
    <AuthProvider>
      <AdminPage defaultTab="inventory" />
    </AuthProvider>
  );
  await waitFor(() => expect(screen.queryByText(/loading/i)).not.toBeInTheDocument(), { timeout: 4000 });
  // Ensure inventory tab is selected
  const inventoryTab = await screen.findByRole('tab', { name: /inventory/i }, { timeout: 8000 });
  fireEvent.click(inventoryTab);
  // Wait for inventory table
  let table;
  try {
    table = await screen.findByRole('table', {}, { timeout: 8000 });
  } catch {
    screen.debug();
    table = screen.queryByRole('table');
  }
  expect(table).not.toBeNull();
  // Try to find the badge by text, fallback to queryAllByText if not found
  let outOfStockBadges;
  try {
    outOfStockBadges = await screen.findAllByText(/out of stock/i, { exact: false }, { timeout: 8000 });
  } catch {
    outOfStockBadges = screen.queryAllByText(/out of stock/i, { exact: false });
  }
  expect(outOfStockBadges.length).toBeGreaterThan(0);
}, 20000);

// Bulk delete dialog
it('shows bulk delete dialog when products are selected', async () => {
  jest.spyOn(require('../lib/firestore-service'), 'getProducts').mockResolvedValueOnce([
    {
      id: 'p1',
      name: 'Test',
      imageUrl: '',
      price: 100,
      stock: 1,
      category: 'Electronics',
      brand: '',
      subcategory: '',
      isFeatured: false,
      createdAt: new Date().toISOString(),
      checked: false,
      description: 'Test product',
      updatedAt: new Date().toISOString(),
      status: 'active'
    },
    {
      id: 'p2',
      name: 'Test 2',
      imageUrl: '',
      price: 200,
      stock: 5,
      category: 'Electronics',
      brand: '',
      subcategory: '',
      isFeatured: false,
      createdAt: new Date().toISOString(),
      checked: false,
      description: 'Test product 2',
      updatedAt: new Date().toISOString(),
      status: 'active'
    }
  ]);
  jest.spyOn(require('../lib/firestore-service'), 'getOrders').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getUsers').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getAuditLogs').mockResolvedValueOnce([]);
  customRender(
    <AuthProvider>
      <AdminPage defaultTab="inventory" />
    </AuthProvider>
  );
  await waitFor(() => expect(screen.queryByText(/loading/i)).not.toBeInTheDocument(), { timeout: 10000 });
  const inventoryTab = screen.getByRole('tab', { name: /inventory/i });
  fireEvent.click(inventoryTab);
  await waitFor(() => {
    expect(inventoryTab).toHaveAttribute('aria-selected', 'true');
  }, { timeout: 10000 });
  let table;
  try {
    table = await screen.findByRole('table', {}, { timeout: 10000 });
  } catch {
    screen.debug();
    table = screen.queryByRole('table');
  }
  expect(table).not.toBeNull();
  let checkboxes;
  try {
    checkboxes = await screen.findAllByRole('checkbox', {}, { timeout: 10000 });
  } catch (e) {
    screen.debug();
    throw e;
  }
  fireEvent.click(checkboxes[1]);
  const deleteBtn = screen.queryByText(/delete \(1\)/i);
  expect(deleteBtn).not.toBeNull();
  expect(deleteBtn).toBeInTheDocument();
}, 20000);

// Add product dialog
it('opens add product dialog', async () => {
  jest.spyOn(require('../lib/firestore-service'), 'getProducts').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getOrders').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getUsers').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getAuditLogs').mockResolvedValueOnce([]);
  customRender(
    <AuthProvider>
      <AdminPage defaultTab="inventory" />
    </AuthProvider>
  );
  await waitFor(() => expect(screen.queryByText(/loading/i)).not.toBeInTheDocument(), { timeout: 4000 });
  let addBtn;
  try {
    addBtn = await screen.findByRole('button', { name: /add new product/i });
  } catch (e) {
    screen.debug();
    throw e;
  }
  fireEvent.click(addBtn);
  expect(await screen.findByText(/add new product/i)).toBeInTheDocument();
});

// Edit product dialog
it('opens edit product dialog', async () => {
  jest.spyOn(require('../lib/firestore-service'), 'getProducts').mockResolvedValueOnce([
    {
      id: 'p1',
      name: 'Test',
      imageUrl: '',
      price: 100,
      stock: 1,
      category: 'Electronics',
      brand: '',
      subcategory: '',
      isFeatured: false,
      createdAt: new Date().toISOString(),
      checked: false,
      description: 'Test product',
      updatedAt: new Date().toISOString(),
      status: 'active'
    }
  ]);
  jest.spyOn(require('../lib/firestore-service'), 'getOrders').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getUsers').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getAuditLogs').mockResolvedValueOnce([]);
  customRender(
    <AuthProvider>
      <AdminPage defaultTab="inventory" />
    </AuthProvider>
  );
  await waitFor(() => expect(screen.queryByText(/loading/i)).not.toBeInTheDocument(), { timeout: 10000 });
  const inventoryTab = screen.getByRole('tab', { name: /inventory/i });
  fireEvent.click(inventoryTab);
  // Wait for the table to appear before proceeding
  let table;
  try {
    table = await screen.findByRole('table', {}, { timeout: 10000 });
  } catch {
    screen.debug();
    table = screen.queryByRole('table');
  }
  expect(table).not.toBeNull();
  let editBtns;
  try {
    editBtns = await screen.findAllByRole('button', { name: /edit/i }, { timeout: 10000 });
  } catch (e) {
    screen.debug();
    throw e;
  }
  fireEvent.click(editBtns[0]);
  expect(await screen.findByText(/edit product/i)).toBeInTheDocument();
}, 20000);

// Error handling: API failure
it('shows error toast on API failure', async () => {
  jest.spyOn(require('../lib/firestore-service'), 'deleteProduct').mockResolvedValueOnce({ success: false, message: 'Failed' });
  jest.spyOn(require('../lib/firestore-service'), 'getProducts').mockResolvedValueOnce([
    {
      id: 'p1',
      name: 'Test',
      imageUrl: '',
      price: 100,
      stock: 1,
      category: 'Electronics',
      brand: '',
      subcategory: '',
      isFeatured: false,
      createdAt: new Date().toISOString(),
      checked: false,
      description: 'Test product',
      updatedAt: new Date().toISOString(),
      status: 'active'
    }
  ]);
  jest.spyOn(require('../lib/firestore-service'), 'getOrders').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getUsers').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getAuditLogs').mockResolvedValueOnce([]);
  customRender(
    <AuthProvider>
      <AdminPage defaultTab="inventory" />
    </AuthProvider>
  );
  await waitFor(() => expect(screen.queryByText(/loading/i)).not.toBeInTheDocument(), { timeout: 10000 });
  const inventoryTab = screen.getByRole('tab', { name: /inventory/i });
  fireEvent.click(inventoryTab);
  let table;
  try {
    table = await screen.findByRole('table', {}, { timeout: 10000 });
  } catch {
    screen.debug();
    table = screen.queryByRole('table');
  }
  expect(table).not.toBeNull();
  let checkboxes;
  try {
    checkboxes = await screen.findAllByRole('checkbox', {}, { timeout: 10000 });
  } catch (e) {
    screen.debug();
    throw e;
  }
  window.confirm = jest.fn(() => true);
  fireEvent.click(checkboxes[1]);
  let deleteBtns;
  try {
    deleteBtns = await screen.findAllByRole('button', { name: /delete/i }, { timeout: 10000 });
  } catch (e) {
    screen.debug();
    throw e;
  }
  window.confirm = jest.fn(() => true);
  await act(async () => {
    fireEvent.click(deleteBtns[0]);
  });
  const errorToast = await screen.findByText(/error/i, {}, { timeout: 10000 });
  expect(errorToast).not.toBeNull();
  expect(errorToast).toBeInTheDocument();
}, 20000);

// Accessibility: heading
it('has accessible heading for manage products', async () => {
  customRender(
    <AuthProvider>
      <AdminPage defaultTab="inventory" />
    </AuthProvider>
  );
  await waitFor(() => expect(screen.queryByText(/loading/i)).not.toBeInTheDocument(), { timeout: 4000 });
  jest.spyOn(require('../lib/firestore-service'), 'getProducts').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getOrders').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getUsers').mockResolvedValueOnce([]);
  jest.spyOn(require('../lib/firestore-service'), 'getAuditLogs').mockResolvedValueOnce([]);
  await waitFor(() => expect(screen.queryByText(/loading/i)).not.toBeInTheDocument(), { timeout: 4000 });
  try {
    expect(await screen.findByRole('heading', { name: /manage products/i })).toBeInTheDocument();
  } catch (e) {
    screen.debug();
    throw e;
  }
});

// Role-based access
// it('shows restriction for non-admin users', async () => {
//   jest.resetModules();
//   jest.doMock('../context/auth-context', () => ({
//     useAuth: () => ({ user: { uid: 'user1', email: 'user@example.com' }, loading: false, role: 'client' }),
//     AuthProvider: ({ children }: any) => <>{children}</>,
//   }));
//   jest.doMock('../lib/firestore-service', () => ({
//     getProducts: jest.fn(() => Promise.resolve([])),
//     getOrders: jest.fn(() => Promise.resolve([])),
//     getUsers: jest.fn(() => Promise.resolve([])),
//     getAuditLogs: jest.fn(() => Promise.resolve([])),
//     deleteProduct: jest.fn(() => Promise.resolve({ success: true })),
//     deleteMultipleProducts: jest.fn(() => Promise.resolve({ success: true })),
//     addProduct: jest.fn(() => Promise.resolve({ success: true })),
//     updateProduct: jest.fn(() => Promise.resolve({ success: true })),
//     updateOrderStatus: jest.fn(() => Promise.resolve({ success: true })),
//     deleteUser: jest.fn(() => Promise.resolve({ success: true })),
//     updateUserRole: jest.fn(() => Promise.resolve({ success: true })),
//   }));
//   const { default: AdminPage } = require('../app/admin/page');
//   const { AuthProvider } = require('../context/auth-context');
//   render(
//     <AuthProvider>
//       <AdminPage defaultTab="inventory" />
//     </AuthProvider>
//   );
//   await waitFor(() => expect(screen.queryByText(/loading/i)).not.toBeInTheDocument(), { timeout: 4000 });
//   try {
//     expect(screen.getByText(/access denied|not authorized/i)).toBeInTheDocument();
//   } catch (e) {
//     screen.debug();
//     throw e;
//   }
// });

// Real-time update simulation
it('updates UI on WebSocket message', async () => {
  // This would require mocking WebSocket and triggering a message
  // For brevity, just check that fetchData is called on message
  // You can use jest.spyOn and simulate ws.onmessage
});

// Snapshot test
it('matches snapshot for inventory tab', async () => {
  const { asFragment } = render(
    <AuthProvider>
      <AdminPage defaultTab="inventory" />
    </AuthProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});
