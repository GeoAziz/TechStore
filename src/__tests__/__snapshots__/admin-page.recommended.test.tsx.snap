// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`matches snapshot for inventory tab 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center h-full"
  >
    <svg
      class="lucide lucide-loader-circle w-12 h-12 animate-spin text-primary"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21 12a9 9 0 1 1-6.219-8.56"
      />
    </svg>
  </div>
</DocumentFragment>
`;
